File: CS 3450 Code Examples/Iterator Stuff/InternalIterator/InternalIter.cpp
#include <iostream>
#include <string>
using namespace std;

template<class Iter, class F>
void forEach(Iter start, Iter stop, F f) {
    while (start != stop)
        f(*start++);
}

template<class T>
void f(const T& t) {
    cout << t << endl;
}

int main() {
    string a[] = {"one", "two", "three"};
    forEach(a, a+3, &f<string>);
}

File: CS 3450 Code Examples/JavaAll/InternalIter.java
import java.util.*;

interface UnaryFunction {
    void execute(Object o);
}

class Display implements UnaryFunction {
    public void execute(Object o) {
        System.out.println(o);
    }
}

class InternalIter {
    public static void forEach(Collection c, UnaryFunction f) {
        Iterator it = c.iterator();
        while (it.hasNext())
            f.execute(it.next());
    }
    public static void main(String[] args) {
        ArrayList stuff = new ArrayList();
        stuff.add("one");
        stuff.add("two");
        stuff.add("three");
        forEach(stuff, new Display());
    }
}
